[BUG]: serious warning in 32bit system

### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Compile-time Error

### Component

General CCCL

### Describe the bug

I am not sure if you care about 32bit systems, but I get this warning deep in CCCL.

```
/builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/__memory/check_address.h:75:52: error: narrowing conversion of '18446744073709551615' from 'long long unsigned int' to 'uintptr_t' {aka 'unsigned int'} [-Wnarrowing]
   75 |   auto __limit = ::cuda::std::uintptr_t{UINTMAX_MAX} - static_cast<::cuda::std::uintptr_t>(__n);
```

When using this compiler

```
current binary directory: /builds/correaa/boost-multi/build
-- The CXX compiler identification is GNU 14.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
```

and this version of CCCL

```
-- Finding CCCL components: Thrust;CUB;libcudacxx
-- Found CUB: /builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/cub/cub-config.cmake (found suitable version "3.3.0.0", minimum required is "3.3.0.0")
-- Found libcudacxx: /builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/libcudacxx/libcudacxx-config.cmake (found suitable exact version "3.3.0.0")
-- Found Thrust: /builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/thrust/thrust-config.cmake (found suitable exact version "3.3.0.0")
-- Found CCCL: /builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/cccl/cccl-config.cmake (found version "3.3.0.0")
-- Found libcudacxx: /builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/libcudacxx/libcudacxx-config.cmake (found version "3.3.0.0")
-- Found CUB: /builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/cub/cub-config.cmake (found version "3.3.0.0")
-- Found Thrust: /builds/correaa/boost-multi/build/_deps/cccl-cmake-src/lib/cmake/thrust/thrust-config.cmake (found version "3.3.0.0")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5") found components: CXX
```

All details are here: https://gitlab.com/correaa/boost-multi/-/jobs/12835122596


### How to Reproduce

All details are here: https://gitlab.com/correaa/boost-multi/-/jobs/12835122596


### Expected behavior

no warnings should be issued

### Reproduction link

https://gitlab.com/correaa/boost-multi/-/jobs/12835122596

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

--- a/libcudacxx/include/cuda/__memory/check_address.h
+++ b/libcudacxx/include/cuda/__memory/check_address.h
@@ -21,10 +21,6 @@
 #  pragma system_header
 #endif // no system header
 
-// including cuda/std/limits generates a circular dependency because:
-//    numeric_limits -> bit_cast -> cstring -> check_address
-// <cuda/std/__utility/cmp.h> also includes cuda/std/limits
-#include <cuda/std/climits>
 #include <cuda/std/cstddef>
 #include <cuda/std/cstdint>
 #if _CCCL_CUDA_COMPILATION()
@@ -72,20 +68,24 @@ _CCCL_BEGIN_NAMESPACE_CUDA
   {
     return false;
   }
-  auto __limit = ::cuda::std::uintptr_t{UINTMAX_MAX} - static_cast<::cuda::std::uintptr_t>(__n);
+
+  // use (~::cuda::std::uintptr_t{0}) instead of cuda::std::numeric_limits<cuda::std::uintptr_t>::max() to avoid
+  // circular dependency because:
+  //    numeric_limits -> bit_cast -> cstring -> check_address
+  // <cuda/std/__utility/cmp.h> also includes cuda/std/limits
+  const auto __limit = (~::cuda::std::uintptr_t{0}) - static_cast<::cuda::std::uintptr_t>(__n);
+
   if (reinterpret_cast<::cuda::std::uintptr_t>(__ptr) > __limit)
   {
     return false;
   }
-  // clang-format off
-  NV_IF_TARGET(NV_IS_DEVICE, (
-    if (::cuda::device::__internal_is_address_from(__ptr, ::cuda::device::address_space::shared) &&
-        !::cuda::device::__is_smem_valid_address_range(__ptr, __n))
-    {
-      return false;
-    }
-  ));
-  // clang-format on
+  NV_IF_TARGET(NV_IS_DEVICE, ({
+                 if (::cuda::device::__internal_is_address_from(__ptr, ::cuda::device::address_space::shared)
+                     && !::cuda::device::__is_smem_valid_address_range(__ptr, __n))
+                 {
+                   return false;
+                 }
+               }));
   return (__ptr != nullptr);
 }
 
@@ -97,10 +97,10 @@ _CCCL_BEGIN_NAMESPACE_CUDA
 [[nodiscard]] _CCCL_API inline bool
 __are_ptrs_overlapping(const void* __ptr_lhs, const void* __ptr_rhs, ::cuda::std::size_t __n) noexcept
 {
-  auto __ptr1_start = static_cast<const char*>(__ptr_lhs);
-  auto __ptr2_start = static_cast<const char*>(__ptr_rhs);
-  auto __ptr1_end   = __ptr1_start + __n;
-  auto __ptr2_end   = __ptr2_start + __n;
+  const auto __ptr1_start = static_cast<const char*>(__ptr_lhs);
+  const auto __ptr2_start = static_cast<const char*>(__ptr_rhs);
+  const auto __ptr1_end   = __ptr1_start + __n;
+  const auto __ptr2_end   = __ptr2_start + __n;
   return __ptr1_start < __ptr2_end && __ptr2_start < __ptr1_end;
 }
 

